core/lib/zksync_core/src/reorg_detector/mod.rs

```diff
@@ -346,9 +356,19 @@ impl ReorgDetector {
             .with_context(|| format!("Root hash does not exist for local batch #{l1_batch}"))?;
         drop(storage);
 
-        let Some(remote_hash) = self.client.l1_batch_root_hash(l1_batch).await? else {
-            tracing::info!("Remote L1 batch #{l1_batch} is missing");
-            return Err(HashMatchError::RemoteHashMissing);
+        let remote_hash = loop {
```

Comment:
how about having this retry loop within l1_batch_root_hash  call? Or how about  treating an l1 batch as nonexistent (for the purpose of reorg detector) until it has root_hash populated?

Reply:
> how about having this retry loop within l1_batch_root_hash call?

This would work, but IMO, that'd break abstraction boundary; it's not client's responsibility to determine the course of action if the batch exists, but its state root hash doesn't. OTOH, the client trait is specific to the reorg detector, so that's a contentious point.

> how about treating an l1 batch as nonexistent (for the purpose of reorg detector) until it has root_hash populated?

Do you mean making multiple calls in `sealed_l1_batch_number()`? Again, could work, but it'd be less efficient, and in a common case when the remote L1 batch is the one to check, the obtained state root hash will be wasted. (Of course, it's possible to change `sealed_l1_batch_number()` signature so that it'd return the state root hash as well, but that looks clunky.)

Reply:
Good catch! IMO, it's reasonable to refactor `check_consistency()` so that it's cancel-aware. Since it does mutate the reorg detector state, this looks preferable to `select!`ing between it and the stop signal future.